File: 03/rb/03.rb
#!/usr/bin/env ruby

require 'set'

def path s
    x, y = 0, 0
    pts = Set.new
    steps = {}
    sc = 0
    s.split(?,).each do |mov|
        n = mov[1..-1].to_i
        xv, yv = case mov[0]
                 when ?U then [ 0,  1]
                 when ?D then [ 0, -1]
                 when ?R then [ 1,  0]
                 when ?L then [-1,  0]
                 end
        (0..n).each do |i|
            pt = [x+xv*i,y+yv*i]
            pts.add pt
            steps[pt] = sc+i unless steps.include? pt
        end
        x += xv*n; y += yv*n; sc += n
    end
    [pts.delete([0,0]), steps]
end

l1, l2 = File.readlines 'input'
p1, s1 = path l1
p2, s2 = path l2

# part 1
p (p1 & p2).map{|x| x.map(&:abs).sum}.min
# part 2
p (p1 & p2).map{|x| s1[x] + s2[x]}.min

File: 03/jl/03.jl
#!/usr/bin/env julia

# input has 300 instructions of length <1000,
# so max coord is 300000
const BASE = Int64(600000)
const DIRS = Dict('U'=>BASE, 'D'=>-BASE, 'R'=>1, 'L'=>-1)

function path(s::String)
    pos = BASE÷2
    pts = Dict{Int64, Int32}()
    step = zero(Int32)

    for x = split(s, ",")
        dir, n = x[1], parse(Int32, x[2:end])
        for _ = 1:n
            pos += DIRS[dir]
            step += 1
            if !haskey(pts, pos); pts[pos] = step; end
        end
    end

    pts
end

d1, d2 = open("input") do f
    map(path, readlines(f))
end

pts = collect(keys(d1) ∩ keys(d2))
# part 1
println(minimum(map(x -> abs(fld(x, BASE)) + abs(mod(x, BASE) - BASE÷2), pts)))
# part 2
println(minimum(map(x -> d1[x] + d2[x], pts)))
